[SPARK-13723] [YARN] Change behavior of --num-executors with dynamic allocation.

rdblue opened the pull request:
## What changes were proposed in this pull request?

This changes the behavior of --num-executors and spark.executor.instances when using dynamic allocation. Instead of turning dynamic allocation off, it uses the value for the initial number of executors.

This changes was discussed on [SPARK-13723](https://issues.apache.org/jira/browse/SPARK-13723). I highly recommend using it while we can change the behavior for 2.0.0. In practice, the 1.x behavior causes unexpected behavior for users (it is not clear that it disables dynamic allocation) and wastes cluster resources because users rarely notice the log message.
## How was this patch tested?

This patch updates tests and adds a test for Utils.getDynamicAllocationInitialExecutors.

rdblue committed:
SPARK-13723: Change behavior of --num-executors with dynamic allocation.

Previously, using --num-executors or spark.executor.instances with
dynamic allocation enabled would disable dynamic allocation. This
updates the behavior so these settings change the initial number of
executors and do not disable dynamic allocation.

SparkQA commented:
**[Test build #59423 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59423/consoleFull)** for PR 13338 at commit [`5f69549`](https://github.com/apache/spark/commit/5f6954943ed9278b3ebe6b871f20b14f66efd2f6).

SparkQA commented:
**[Test build #59423 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59423/consoleFull)** for PR 13338 at commit [`5f69549`](https://github.com/apache/spark/commit/5f6954943ed9278b3ebe6b871f20b14f66efd2f6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59423/
Test PASSed.

jerryshao commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -2262,21 +2262,39 @@ private[spark] object Utils extends Logging {
+  def getDynamicAllocationInitialExecutors(conf: SparkConf): Int = {
+    math.max(
+      math.max(
+        conf.getInt("spark.dynamicAllocation.initialExecutors", 0),
```
Maybe we could change to who take precedence to who when both set, using `max` may not be a good choice.

tgravescs replied:
I'm not overly concerned with this as long as its clear to the users which way we chose.  It feels like the user would use --num-executors more easily or accidentally then specifying spark.dynamicAllocation.initialExecutors.

How about leaving it max and just print a warning if we find both set and tell them what we choose.  

I'm also not sure there is a good way to do this right now with the ConfigEntry/Builder stuff thouh. @vanzin

vanzin replied:
I don't think the config API can handle both calls to `math.max` here... you could replace the inner one with a fallback config (`initialExecutors` falls back to the value of `spark.executor.instances` if not set), although even then the semantics are slightly different.

In any case, this might be more readable as:

```
Seq(
  conf.getInt("spark.dynamicAllocation.initialExecutors", 0),
  conf.getInt("spark.executor.instances", 0),
  getDynamicAllocationMinExecutors(conf)).max
```

Assuming I got the semantics right.

rdblue replied:
I'm not sure what you mean by the config API can't handle both calls to max.

I'm happy to make the change for this to be more readable, but I think we should decide on semantics first. I think a reasonable way to handle this is to use the max of all 3, as implemented. The min value should clearly be handled this way -- starting at initial less than min causes an immediate change to the min -- so the question is how to handle both `spark.executor.instances` and `spark.dynamicAllocation.initialExecutors`.

I think it's reasonable to use either one in different situations so I don't think it makes much sense to complain if they're both set. My job could have initialExecutors set in its config, but I can run it with --num-executors to bump up the value. Given that use case, I don't think it would be a good idea to have initialExecutors override spark.executor.instances.

That leaves whether spark.executor.instances should override initialExecutors or whether the initial number should be the max of the two. I don't have a strong opinion here, but I opted for the max of the two values.

vanzin replied:
> I'm not sure what you mean by the config API can't handle both calls to max.

I was replying to Tom's question, you can ignore that part.

tgravescs replied:
I'm fine with max and as I mentioned I agree that I think users might use or accidentally use --num-executors to bump up the value.  Although I'd rather not encourage that behavior because its intermixing these things and --num-executors is ambiguous on the behavior (based on the dynamic allocation on or off), which is why I suggested the warning. The warning was also in case someone specified --num-executors and thought they were getting static number but instead had dynamic allocation on so its initial.  

I would also be fine with just an info message here printing what allocation method we are using and the number chosen.

jerryshao commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala:
```diff
@@ -519,9 +519,9 @@ object YarnSparkHadoopUtil {
-      val minNumExecutors = conf.get(DYN_ALLOCATION_MIN_EXECUTORS)
-      val initialNumExecutors = conf.get(DYN_ALLOCATION_INITIAL_EXECUTORS)
-      val maxNumExecutors = conf.get(DYN_ALLOCATION_MAX_EXECUTORS)
+      val minNumExecutors = Utils.getDynamicAllocationMinExecutors(conf)
```
It would be better to use `DYN_ALLOCATION_MIN_EXECUTORS` from my understanding.

rdblue replied:
`DYN_ALLOCATION_MIN_EXECUTORS` is a String property. Instead of relying on the constants, this now standardizes on using functions in `Utils`.

vanzin replied:
Hmm, no. `sparkConf.get(DYN_ALLOCATION_MAX_EXECUTORS)` will return you an int, and if it's not defined in the conf, it will return you the default value, so you won't need to duplicate it in your code.

rdblue replied:
Sorry, I thought that DYN_ALLOCATION_MAX_EXECUTORS was a String. Marcelo's comment cleared it up and I've fixed these. Thanks!

tgravescs commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -2262,21 +2262,39 @@ private[spark] object Utils extends Logging {
+   * Return the minimum number of executors for dynamic allocation.
+   */
+  def getDynamicAllocationMinExecutors(conf: SparkConf): Int = {
+    conf.getInt("spark.dynamicAllocation.minExecutors", 0)
```
as stated below we should use DYN_ALLOCATION_MIN_EXECUTORS, DYN_ALLOCATION_INITIAL_EXECUTORS,  and DYN_ALLOCATION_MAX_EXECUTORS

vanzin replied:
yep, don't add these methods. instead:

```
import org.apache.spark.internal.config._

conf.get(DYN_ALLOCATION_MIN_EXECUTORS)
```

Also recommend going through your change and modifying all the places that hardcode the config key to use the constants.

rdblue replied:
What's the value of these constants over util methods? These constants live in the yarn module, but there are uses in the `ExecutorAllocationManager` in core. Rather than having the constants in two places, with possibly different defaults, I thought a util method was appropriate.

vanzin replied:
> What's the value of these constants over util methods?

They encapsulate key, type and default value already, so you'd avoid duplicating those in your code.

> These constants live in the yarn module

No, they're in `core/src/main/scala/org/apache/spark/internal/config/package.scala`.

> Rather than having the constants in two places, with possibly different defaults, I thought a util method was appropriate.

That's exactly what those constants are.

rdblue replied:
I see, so those aren't Strings. Thanks, I'll fix it.

rdblue replied:
Fixed.

tgravescs commented on docs/configuration.md:
```diff
@@ -1158,6 +1158,10 @@ Apart from these, the following properties are also available, and may be useful
     <a href="job-scheduling.html#dynamic-resource-allocation">here</a>.
     <br><br>
+    When this property is enabled, <code>spark.executor.instances</code> will affect the initial
+    number of executors, increasing the initial number if it is higher than the minimum or the
```
I think the wording here is a bit confusing.    Perhaps remove everything after the , and just say see spark.dynamicAllocation.initialExecutors and we can add more description there.

rdblue replied:
I'll update docs after semantics are discussed.

rdblue replied:
Fixed.

tgravescs commented:
I think we should also update the spark-submit --help for --num-executors.

rdblue commented:
Thanks for reviewing, everyone! I've made some comments and will update once we have consensus on util methods and semantics.

SparkQA commented:
**[Test build #60423 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60423/consoleFull)** for PR 13338 at commit [`b27a5b8`](https://github.com/apache/spark/commit/b27a5b85a12950e45dc8391808116104421543a7).

rdblue commented:
@tgravescs, where is the --help text for spark-submit? I'll update it.

rdblue commented:
I just pushed a version that addresses the comments so far (other than the spark-submit help text). Thanks for looking at this, @vanzin, @jerryshao,  and @tgravescs!

SparkQA commented:
**[Test build #60423 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60423/consoleFull)** for PR 13338 at commit [`b27a5b8`](https://github.com/apache/spark/commit/b27a5b85a12950e45dc8391808116104421543a7).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60423/
Test FAILed.

tgravescs commented:
Its in SparkSubmitArguments.scala function printUsageAndExit

rdblue committed:
SPARK-13723: Update --num-executors in spark-submit help.

SparkQA commented:
**[Test build #60426 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60426/consoleFull)** for PR 13338 at commit [`47cff98`](https://github.com/apache/spark/commit/47cff98f765328c2219160ac61aec09573a85dbf).

SparkQA commented:
**[Test build #60427 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60427/consoleFull)** for PR 13338 at commit [`bf22b5a`](https://github.com/apache/spark/commit/bf22b5ab0bc8369949ac33833b078e7e13c7ce35).

jerryshao commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -2309,21 +2310,24 @@ private[spark] object Utils extends Logging {
+    Seq(
+      conf.get(DYN_ALLOCATION_MIN_EXECUTORS),
+      conf.get(DYN_ALLOCATION_INITIAL_EXECUTORS),
+      conf.get(EXECUTOR_INSTANCES).getOrElse(0)).max
```
Do we need to support environment variable `SPARK_EXECUTOR_INSTANCES`? Since it is not officially deprecated.

tgravescs replied:
EXECUTOR_INSTANCES isn't deprecated.  This is a conf (spark.executor.instances), its not documented but I'm pretty sure I've seen people using it instead of --num-executors and that is what SparkSubmit uses.

tgravescs replied:
sorry I was mistaken it is documented on the yarn conf page.

jerryshao replied:
I think what I mean is env variable `SPARK_EXECUTOR_INSTANCES `, we should also consider this variable when getting dynamic allocation initial executors (in case someone is using this variable).

tgravescs replied:
oh I see. Well we didn't support it before in this routine and I would prefer not to add more support to env variables as I would rather see them all removed.  So I would say no, at least not under this jira.  The only place I see it being used is in the YarnSparkHadoopUtil. getInitialTargetExecutorNumber so I would say we file a separate jira to remove it and clean up comments in code in other places.

SparkQA commented:
**[Test build #60427 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60427/consoleFull)** for PR 13338 at commit [`bf22b5a`](https://github.com/apache/spark/commit/bf22b5ab0bc8369949ac33833b078e7e13c7ce35).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60427/
Test FAILed.

SparkQA commented:
**[Test build #60426 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60426/consoleFull)** for PR 13338 at commit [`47cff98`](https://github.com/apache/spark/commit/47cff98f765328c2219160ac61aec09573a85dbf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60426/
Test PASSed.

tgravescs commented:
we need to update running-on-yarn.md the description for spark.executor.instances as it says dynamic allocation will be turned off when its specified.

otherwise LGTM.

tgravescs commented:
ping @rdblue

tgravescs commented:
I really want to get this into the 2.0 release since its a change in behavior if you don't have time I'll put up pr with minor changes.

rdblue commented:
@tgravescs, I've updated it. Sorry about the delay, for some reason the notifications for this issue didn't make it to my inbox so I wasn't seeing updates.

rdblue committed:
SPARK-13723: Update running-on-yarn.md for these changes.

rdblue commented:
@tgravescs, I've updated it. Sorry about the delay, for some reason the notifications for this issue didn't make it to my inbox so I wasn't seeing updates.

SparkQA commented:
**[Test build #61118 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61118/consoleFull)** for PR 13338 at commit [`0352883`](https://github.com/apache/spark/commit/0352883dba5aef4fd56d6aaddba129071b1d22ea).

SparkQA commented:
**[Test build #61118 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61118/consoleFull)** for PR 13338 at commit [`0352883`](https://github.com/apache/spark/commit/0352883dba5aef4fd56d6aaddba129071b1d22ea).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61118/
Test PASSed.

tgravescs commented:
+1, thanks @rdblue

rdblue commented:
Thanks @tgravescs!
